### PR TITLE
TOML declarative config for modules 4/4

### PR DESCRIPTION
**Pull request by janciesla8818**
Modules covered in this PR:
- [x] mod_register
- [x] mod_roster
- [x] mod_shared_roster_ldap
- [x] mod_stream_management
- [x] mod_vcard
- [x] mod_version
- [x] mod_revproxy - deprecated, TOML parsing and validation removed
- [x] mod_last 

Checklist for each module:
- Write the new spec in the module as a callback function: `config_spec/0`. Some cases need reworking the options, e.g. when the same value can be a primitive (e.g. string or atom) or a nested section/list
- Add the module to `mongoose_config_spec:all_modules/0` and `HAS_NO_SPEC`
- Double check with the docs: `https://esl.github.io/MongooseDocs` that the options and their format, validator etc are correct
- Remove `module_opt` and `handler` clauses from `mongoose_config_parser_toml` as well as module-specific functions for nested options
- Remove `validate` clauses from `mongoose_config_validator_toml`
- Make sure this passes locally: `./rebar3 ct --suite=config_parser_SUITE` before pushing anything, also check `./rebar3 dialyzer`
- Check test coverage after pushing, there might be some options not covered by tests, you can do it by looking at the test cases as well
- Regarding tests, it's good to simplify them and fix indentation/maximum line length of 100 characters, see the change for `mod_adhoc` or `mod_auth_token`.
- Sanity check: grep for the module name in `src/config`

### CODE REVIEW

**Review by chrzaszcz — COMMENTED** [review]
Reviewed `mod_register`: a few minor comments. It's good, but there are a few easy quick wins to improve it.

File: src/mod_roster.erl [hunk 1]
```diff
@@ -212,6 +214,26 @@ stop(Host) ->
     ejabberd_hooks:delete(hooks(Host)),
     gen_iq_handler:remove_iq_handler(ejabberd_sm, Host, ?NS_ROSTER).
 
+-spec config_spec() -> mongoose_config_spec:config_section().
+config_spec() ->
+    #section{
+       items = #{<<"iqdisc">> => mongoose_config_spec:iqdisc(),
+                <<"versioning">> => #option{type = boolean},
```
Comment (chrzaszcz) [inline src/mod_roster.erl:221]: Minor: indentation

File: test/config_parser_SUITE.erl [hunk 2]
```diff
@@ -2557,29 +2557,28 @@ mod_revproxy(_Config) ->
...
-            ?_errf(T(#{<<"store_current_id">> => 1})),
-            ?_errf(T(#{<<"backend">> => 1})),
-            ?_errf(T(#{<<"backend">> => <<"iloveyou">>})),
-            ?_errf(T(#{<<"riak">> => #{<<"version_bucket_type">> => 1}})),
-            ?_errf(T(#{<<"riak">> => #{<<"bucket_type">> => 1}}))
-          ]),
+    M = fun(Cfg) -> modopts(mod_roster, Cfg) end,
+
+    ?eqf(M([{iqdisc, one_queue}]),
+       T(#{<<"iqdisc">> => #{<<"type">> => <<"one_queue">>}})),
```
Comment (chrzaszcz) [inline test/config_parser_SUITE.erl:2542]: Do we need it if we already do `check_iqdisc`?

File: test/config_parser_SUITE.erl [hunk 3]
```diff
@@ -2557,29 +2557,28 @@ mod_revproxy(_Config) ->
...
+    ?eqf(M([{store_current_id, false}]),
+       T(#{<<"store_current_id">> => false})),
+    ?eqf(M([{versioning, false}]),
+       T(#{<<"versioning">> => false})),
```
Comment (chrzaszcz) [inline test/config_parser_SUITE.erl:2546]: These two could be swapped as their order in the code and docs is reversed.

**Review by chrzaszcz — COMMENTED** [review]
Reviewed `mod_roster`: all good, just a few tiny remarks.

**Review by chrzaszcz — COMMENTED** [review]
Reviewed `mod_shared_roster_ldap` and liked it a lot, nothing to add there!

File: src/config/mongoose_config_validator_toml.erl [hunk 4]
```diff
@@ -519,6 +501,10 @@ validate(V, integer, positive) -> validate_positive_integer(V);
 validate(V, integer, port) -> validate_port(V);
 validate(V, int_or_infinity, non_negative) -> validate_non_negative_integer_or_infinity(V);
 validate(V, int_or_infinity, positive) -> validate_positive_integer_or_infinity(V);
+validate(V, int_or_infinity_or_atom, positive) ->
+    validate_positive_integer_or_infinity_or_atom(V, no_buffer);
+validate(V, int_or_atom, positive) ->
+    validate_positive_integer_or_atom(V, never);
```
Comment (chrzaszcz) [inline src/config/mongoose_config_validator_toml.erl:447]: We need to think about how to get rid of these strange types... I'd do something like:

- `buffer`: boolean (`disabled/enabled`)
- `buffer_max`: number/infinity

- `ack`: boolean (`disabled/enabled`)
- `ack_freq`: number

It's not mandatory, we could do it later, it just looks confusing and inconsistent as it is now.

File: src/mod_stream_management.erl [hunk 5]
```diff
@@ -73,6 +75,31 @@ stop(Host) ->
         true -> stream_management_stale_h:stop()
     end.
 
+-spec config_spec() -> mongoose_config_spec:config_section().
+config_spec() ->
+    #section{
+       items = #{<<"buffer_max">> => #option{type = int_or_infinity_or_atom,
```
Comment (chrzaszcz) [inline src/mod_stream_management.erl:81]: This type makes me sad...

File: src/mod_stream_management.erl [hunk 6]
```diff
@@ -73,6 +75,31 @@ stop(Host) ->
...
+                                             validate = positive},
+                 <<"ack_freq">> => #option{type = int_or_atom,
```
Comment (chrzaszcz) [inline src/mod_stream_management.erl:83]: This too... it was used only once before and I was planning to get rid of it.

**Review by chrzaszcz — COMMENTED** [review]
Reviewed `mod_stream_management` - the code is good, but I think the TOML types may benefit from disambiguation.

File: src/mod_vcard.erl [hunk 7]
```diff
@@ -196,6 +204,114 @@ stop(VHost) ->
...
+%%--------------------------------------------------------------------
+%% config_spec
+%%--------------------------------------------------------------------
+
+-spec config_spec() -> mongoose_config_spec:config_section().
+config_spec() ->
+    #section{
+       items = #{<<"iqdisc">> => mongoose_config_spec:iqdisc(),
+                 <<"host">> => #option{type = string,
+                                       validate = non_empty},
```
Comment (chrzaszcz) [inline src/mod_vcard.erl:216]: Validate `domain_template` here

File: src/mod_vcard.erl [hunk 8]
```diff
@@ -196,6 +204,114 @@ stop(VHost) ->
...
+                 <<"search">> => #option{type = boolean},
+                 <<"backend">> => #option{type = atom,
+                                          validate = {module, mod_vcard}},
+                 <<"matches">> => #option{type = int_or_infinity,
+                                          validate = non_negative},
+                 <<"ldap_pool_tag">> => #option{type = atom,
+                                                validate = non_empty},
```
Comment (chrzaszcz) [inline src/mod_vcard.erl:223]: Validate `pool_name` here

File: src/mod_vcard.erl [hunk 9]
```diff
@@ -196,6 +204,114 @@ stop(VHost) ->
...
+                 <<"ldap_base">> => #option{type = string},
+                 <<"ldap_uids">> => #list{items = ldap_uids_spec()},
```
Comment (chrzaszcz) [inline src/mod_vcard.erl:225]: This seems identical to `mongoose_config_spec:ldap_uids/0`, maybe reuse that?

File: src/mod_vcard.erl [hunk 10]
```diff
@@ -196,6 +204,114 @@ stop(VHost) ->
...
+                 <<"ldap_binary_search_fields">> => #list{items = #option{type = binary,
+                                                                         validate = non_empty}},
+                 <<"riak">> => riak_config_spec()
+                }
+      }.
+
+ldap_uids_spec() ->
+    #section{
+        items = #{<<"attr">> => #option{type = string},
+                <<"format">> => #option{type = string}},
```
Comment (chrzaszcz) [inline src/mod_vcard.erl:243]: Minor: indentation

File: src/mod_vcard.erl [hunk 11]
```diff
@@ -196,6 +204,114 @@ stop(VHost) ->
...
+                 <<"ldap_filter">> => #option{type = string},
+                 <<"ldap_deref">> => #option{type = atom,
+                                             validate = {enum, [never, always, finding, searching]}},
+                 <<"ldap_vcard_map">> => #list{items = ldap_vcard_map_spec()},
+                 <<"ldap_search_fields">> => #list{items = ldap_search_fields_spec()},
+                 <<"ldap_search_reported">> => #list{items = ldap_search_reported_spec()},
+                 <<"ldap_search_operator">> => #option{type = atom,
+                                                       validate = {enum, ['or', 'and']}},
+                 <<"ldap_binary_search_fields">> => #list{items = #option{type = binary,
+                                                                         validate = non_empty}},
```
Comment (chrzaszcz) [inline src/mod_vcard.erl:235]: Minor: indent

File: src/mod_vcard.erl [hunk 12]
```diff
@@ -196,6 +204,114 @@ stop(VHost) ->
...
+                                              validate = non_empty}
+                },
+        required = all,
+        process = fun ?MODULE:process_search_spec/1
+    }.
+
+ldap_search_reported_spec() ->
+    #section{
+        items = #{<<"search_field">> => #option{type = binary,
+                                               validate = non_empty},
```
Comment (chrzaszcz) [inline src/mod_vcard.erl:282]: Minor: indent

File: test/config_parser_SUITE.erl [hunk 13]
```diff
@@ -2679,91 +2683,109 @@ mod_time(_Config) ->
...
+    [?errf(T([maps:remove(Key, RequiredOpts)])) || Key <- maps:keys(RequiredOpts)],
+    ?errf(T(RequiredOpts#{<<"attr">> := 1})),
+    ?errf(T(RequiredOpts#{<<"format">> => true})).
+
+mod_vcard_ldap_vcard_map(_Config) ->
+    T = fun(Opts) -> #{<<"modules">> =>
+                        #{<<"mod_vcard">> => #{<<"ldap_vcard_map">> => Opts}}} end,
+    M = fun(Cfg) -> modopts(mod_vcard, [{ldap_vcard_map, Cfg}]) end,
+    RequiredOpts = #{<<"vcard_field">> => <<"FAMILY">>,
+                      <<"ldap_pattern">> => <<"%s">>,
```
Comment (chrzaszcz) [inline test/config_parser_SUITE.erl:2727]: Minor: indent, occurs twice below as well.

**Review by chrzaszcz — COMMENTED** [review]
Reviewed `mod_vcard`: quite big, nice improvement of the tests! Just a few remarks from me.

File: src/mod_version.erl [hunk 14]
```diff
@@ -5,8 +5,9 @@
 
 -include("jlib.hrl").
 -include("mongoose.hrl").
+-include("mongoose_config_spec.hrl").
 
--export([start/2, stop/1, process_iq/4]).
+-export([start/2, stop/1, config_spec/0,process_iq/4]).
```
Comment (chrzaszcz) [inline src/mod_version.erl:10]: Minor: space missing

File: src/mod_version.erl [hunk 15]
```diff
@@ -22,6 +23,14 @@ stop(Host) ->
     mod_disco:unregister_feature(Host, ?NS_VERSION),
     gen_iq_handler:remove_iq_handler(ejabberd_local, Host, ?NS_VERSION).
 
+-spec config_spec() -> mongoose_config_spec:config_section().
+config_spec() ->
+    #section{
+       items = #{<<"iqdisc">> => mongoose_config_spec:iqdisc(),
```
Comment (chrzaszcz) [inline src/mod_version.erl:29]: The docs say: `Default: "no_queue"`
The code says: `IQDisc = gen_mod:get_opt(iqdisc, Opts, one_queue),`
Could you fix the docs?

**Review by chrzaszcz — COMMENTED** [review]
Good! A minor remark about making it even better ;-)

**Comment by chrzaszcz** [comment]
Reviewed all, code looks good! Some minor improvements would be nice, see comments, which are grouped per module.

File: src/mod_register.erl [hunk 16]
```diff
@@ -65,6 +69,50 @@ stop(Host) ->
     gen_iq_handler:remove_iq_handler(ejabberd_local, Host, ?NS_REGISTER),
     gen_iq_handler:remove_iq_handler(ejabberd_sm, Host, ?NS_REGISTER).
 
+-spec config_spec() -> mongoose_config_spec:config_section().
+config_spec() ->
+    #section{
+       items = #{<<"iqdisc">> => mongoose_config_spec:iqdisc(),
+                 <<"access">> => #option{type = atom,
+                                         validate = non_empty},
```
[the remark this answers is not in the document]
Comment (janciesla8818) [reply src/mod_register.erl:77]: true, I will change that

File: src/mod_register.erl [hunk 17]
```diff
@@ -65,6 +69,50 @@ stop(Host) ->
...
+
+process_ip_access(KVs) ->
+    {[[{address, Address}], [{policy, Policy}]], []} = proplists:split(KVs, [address, policy]),
+    {Policy, Address}.
+
+process_welcome_message(KVs) ->
+    Body = proplists:get_value(body, KVs, ""),
+    Subject = proplists:get_value(subject, KVs, ""),
+    {Subject, Body}.
+
```
[the remark this answers is not in the document]
Comment (janciesla8818) [reply src/mod_register.erl:115]: The enemy will be surely confused! I'll add it as an item in the cleanup task.

File: test/config_parser_SUITE.erl (hunk 2 above)
Comment (janciesla8818) [reply test/config_parser_SUITE.erl:2542]: True, I'll have that removed and more `check_iqdisc` to the top as it's the first option in the doc.

File: src/mod_vcard.erl (hunk 8 above)
Comment (janciesla8818) [reply src/mod_vcard.erl:223]: I took those ldap types and validations from [ldap_auth](https://github.com/esl/MongooseIM/blob/37f000962df086f5eebeb718dc60fc9e41dd43d0/src/config/mongoose_config_spec.erl#L444), maybe we should change there as well?

File: src/mod_vcard.erl (hunk 9 above)
Comment (janciesla8818) [reply src/mod_vcard.erl:225]: good point

File: src/mod_vcard.erl (hunk 10 above)
Comment (janciesla8818) [reply src/mod_vcard.erl:243]: Will reuse the one from `mongoose_config_spec:ldap_uids()` and have removed `ldap_uids_spec()` from `mod_vcard`.

File: test/config_parser_SUITE.erl (hunk 13 above)
Comment (janciesla8818) [reply test/config_parser_SUITE.erl:2727]: good catch!

File: test/config_parser_SUITE.erl [hunk 18]
```diff
@@ -1870,17 +1870,16 @@ mod_keystore(_Config) ->
     ?errf(T(Base#{<<"keys">> => [InvalidTypeKey]})).
 
 mod_last(_Config) ->
+    check_iqdisc(mod_roster),
```
Comment (chrzaszcz) [inline test/config_parser_SUITE.erl:1873]: mod_last, not roster please

File: src/mod_vcard.erl (hunk 8 above)
Comment (chrzaszcz) [reply src/mod_vcard.erl]: Probably yes, but out of scope here.

**Review by chrzaszcz — APPROVED** [review]
Good work!
Remaining changes regarding types in `mod_string_management` should be addressed in a separate PR.

**Pull request merged**